### PR TITLE
feat: create model from GitHub

### LINKED ISSUE #47: Support multiple image inference for multipart form upload

### LINKED ISSUE #54: Adopt koanf for configuration

### LINKED ISSUE #55: Align database naming with all backend services
The current database table and column names are not consistent across different backend services.

### LINKED ISSUE #62: Read `task` from README.md of model files
No matter uploading a local model or importing a model from GitHub, we extract the `task` field from the `README.md` front matter. The' task' field will not be set if there is no `README.md` or `task` front matter in the `README.md` in the model files.

The `task` of a model is determined by the first version. All future versions of the same model must be with the same `task`.

### LINKED ISSUE #63: Parse `description` field from GitHub repo
The `description` (optional) field is a short description of the model. It won't support Markdown.
- If importing a model from a GitHub repo, `description` is parsed from the `description` of the GitHub repo.
- If uploading a model locally, `description` is provided by the user.

### LINKED ISSUE #67: Model file still exist event creating a model got error
Create a model get error such as task is not consistent with the previously created model. The model weight files still exist.
Need delete these files in failure cases

### PR DESCRIPTION
**Pull request by Phelan164**
Because

- user wanna import models from their public GitHub repository

This commit

- add the method of creating a model from GitHub
- enhance Makefile, using `make dev` to set up local db to work with `air`
- close #55 
- close #54
- close #47 
- close #67 
- close #63
- close #62
- close #54

### CODE REVIEW

File: internal/db/migration/000005_rename_to_singular.up.sql [hunk 1]
```diff
@@ -0,0 +1,7 @@
+BEGIN;
+
+ALTER TABLE "models" RENAME TO "model";
```
Comment (pinglin) [inline internal/db/migration/000005_rename_to_singular.up.sql:3]: Besides changing the table name, we also need to change the constraint name such as
```
ALTER TABLE model RENAME CONSTRAINT models_pkey TO model_pkey
```

File: internal/db/migration/000005_rename_to_singular.up.sql [hunk 2]
```diff
@@ -0,0 +1,7 @@
...
+ALTER TABLE "versions" RENAME TO "version";
```
Comment (pinglin) [inline internal/db/migration/000005_rename_to_singular.up.sql:4]: Besides changing the table name, we also need to change the constraint name:
```
ALTER TABLE version RENAME CONSTRAINT versions_pkey TO version_pkey
```

File: internal/db/migration/000005_rename_to_singular.up.sql [hunk 3]
```diff
@@ -0,0 +1,7 @@
...
+ALTER TABLE "t_models" RENAME TO "triton_model";
```
Comment (pinglin) [inline internal/db/migration/000005_rename_to_singular.up.sql:5]: Besides changing the table name, we also need to change the constraint name:
```
ALTER TABLE triton_model RENAME CONSTRAINT t_models_pkey TO triton_models_pkey
```
Also the foreign-key constraint name:
```
ALTER TABLE triton_model RENAME CONSTRAINT fk_tmodel_version TO fk_triton_model_version
```

**Comment by pinglin** [comment]
Shouldn't we use `JSONB` type instead?

https://github.com/instill-ai/model-backend/blob/e1fb615cf0c06544de30865ad02522f5a9271ddb/internal/db/migration/000001_init.up.sql#L21

Also a typo

https://github.com/instill-ai/model-backend/blob/e1fb615cf0c06544de30865ad02522f5a9271ddb/internal/db/migration/000001_init.up.sql#L12

`namespace in which the model belongs to`

**Comment by pinglin** [comment]
To use singular table name, GORM requires additional [configuration](https://gorm.io/docs/gorm_config.html#NamingStrategy):
```go
db, err := gorm.Open(postgres.Open(connStr), &gorm.Config{
    NamingStrategy: schema.NamingStrategy{
        SingularTable: true,
    },
})
```

**Comment by pinglin** [comment]
Two more issues to consider in this PR:
- Let's adopt [gorm.Model](https://gorm.io/docs/models.html#gorm-Model) in [type Model struct](https://github.com/instill-ai/model-backend/blob/e1fb615cf0c06544de30865ad02522f5a9271ddb/pkg/datamodel/datamodel.go#L10).

- I think we should also make an explicit mapping for the enum type `valid_status` with [gorm](https://github.com/go-gorm/gorm/issues/1978#issuecomment-476673540).

File: internal/db/migration/000005_rename_to_singular.up.sql (hunk 3 above)
Comment (Phelan164) [reply internal/db/migration/000005_rename_to_singular.up.sql:5]: @pinglin yes I just keep origin name for simpler. But okie let change it for more consistency

Comment (Phelan164) [reply internal/db/migration/000005_rename_to_singular.up.sql:5]: btw, beside constraint name we need rename index also. What do you think @pinglin ?

**Comment by Phelan164** [comment]
@pinglin 
>To use singular table name, GORM requires additional [configuration (https://gorm.io/docs/gorm_config.html#NamingStrategy):
I overwrite TableName explicitly then no need setting that. But I will update the setting.

**Comment by Phelan164** [comment]
>Let's adopt [gorm.Model](https://gorm.io/docs/models.html#gorm-Model) in [type Model struct](https://github.com/instill-ai/model-backend/blob/e1fb615cf0c06544de30865ad02522f5a9271ddb/pkg/datamodel/datamodel.go#L10).

model-backend DB has tables model and version
model table do not containcreated_at and updated_at. Because each time creates/update a model, we create/update a corresponding model version, so I think only keep created_at and updated_at in the version. So this model no need to adopt gorm.Model
version table have pair (model_id and version) instead of Id. So it is also not suitable to adopt gorm.Model
What do you think?

**Comment by Phelan164** [comment]
>I think we should also make an explicit mapping for the enum type valid_status with https://github.com/go-gorm/gorm/issues/1978#issuecomment-476673540.

yes, I agree

File: pkg/service/service.go [hunk 4]
```diff
@@ -439,39 +458,39 @@ func (s *service) DeleteModel(namespace string, modelName string) error {
...
 		return err
 	}
 
-	if err := setModelOffline(s, modelInDB.Id, modelVersionInDB.Version); err != nil {
+	if err := setModelOffline(s, modelInDB.ID, modelVersionInDB.Version); err != nil {
 		return err
 	}
 
-	tritonModels, err := s.repository.GetTModelVersions(modelInDB.Id, modelVersionInDB.Version)
+	tritonModels, err := s.repository.GetTModelVersions(modelInDB.ID, modelVersionInDB.Version)
```
Comment (pinglin) [inline pkg/service/service.go:478]:
```suggestion
	tritonModels, err := s.repository.GetTritonModelVersions(modelInDB.ID, modelVersionInDB.Version)
```

File: pkg/datamodel/datamodel.go [hunk 5]
```diff
@@ -42,34 +52,46 @@ type TModel struct {
...
+func (Model) TableName() string {
+	return "model"
+}
+
+func (Version) TableName() string {
+	return "version"
+}
+
+func (TritonModel) TableName() string {
+	return "triton_model"
```
Comment (pinglin) [inline pkg/datamodel/datamodel.go:94]:
```suggestion
```

File: pkg/datamodel/datamodel.go [hunk 6]
```diff
@@ -86,6 +108,79 @@ func (j *JSONB) Scan(value interface{}) error {
...
+		return err
+	}
+
+	return nil
+}
+
+func (r *TritonModel) Value() (driver.Value, error) {
+	valueString, err := json.Marshal(r)
+	return string(valueString), err
+}
```
Comment (pinglin) [inline pkg/datamodel/datamodel.go:168]: Just wondering does `triton_model` have a JSONB column to marshal/unmarshal?

**Review by pinglin — CHANGES_REQUESTED** [review]

**Review by pinglin — DISMISSED** [review]

File: internal/util/util.go [hunk 7]
```diff
@@ -1,11 +1,153 @@
...
+	"path/filepath"
+	"strings"
+
+	"github.com/gernest/front"
+	"github.com/instill-ai/model-backend/pkg/datamodel"
+	"github.com/mitchellh/mapstructure"
+	"gopkg.in/src-d/go-git.v4"
+	"gopkg.in/src-d/go-git.v4/config"
+	"gopkg.in/src-d/go-git.v4/plumbing"
+)
```
Comment (pinglin) [inline internal/util/util.go:20]:
```suggestion
	"github.com/gernest/front"
	"github.com/mitchellh/mapstructure"
	"gopkg.in/src-d/go-git.v4"
	"gopkg.in/src-d/go-git.v4/config"
	"gopkg.in/src-d/go-git.v4/plumbing"
	
	"github.com/instill-ai/model-backend/pkg/datamodel"
)
```

File: pkg/service/service_test.go [hunk 8]
```diff
@@ -6,14 +6,14 @@ package service_test
...
-	"time"
 
 	"github.com/gogo/status"
 	"github.com/golang/mock/gomock"
 	"github.com/stretchr/testify/assert"
 	"google.golang.org/grpc/codes"
+	"gorm.io/gorm"
 
-	"github.com/instill-ai/model-backend/internal/inferenceserver"
+	inferenceserver "github.com/instill-ai/model-backend/internal/inferenceserver"
```
Comment (pinglin) [inline pkg/service/service_test.go:16]:
```suggestion
	"github.com/instill-ai/model-backend/internal/inferenceserver"
```

**Review by pinglin — CHANGES_REQUESTED** [review]
The last few minor suggested changes. Unit and integration tests have passed.

**Review by pinglin — APPROVED** [review]

**Review by pinglin — DISMISSED** [review]
LGTM

**Review by xiaofei-du — APPROVED** [review]

**Pull request merged**